Tiered storage tests

Hangleton opened the pull request.

Hangleton committed:
Add an option to deactivate actual delete in the LocalRemoteStorageManager. Eases testing by allowing simulation of non-strongly consistent storage systems which do not guarantee visibility of a successful delete for subsequent read or list operations.

Hangleton committed:
Add an option to deactivate actual delete in the LocalRemoteStorageManager. Eases testing by allowing simulation of non-strongly consistent storage systems which do not guarantee visibility of a successful delete for subsequent read or list operations.

Hangleton committed:
Add an option to deactivate actual delete in the LocalRemoteStorageManager. Eases testing by allowing simulation of non-strongly consistent storage systems which do not guarantee visibility of a successful delete for subsequent read or list operations.

Hangleton committed:
- Add visitor to traverse locally emulated remote storage to provide support for test assertions.
- Pass connectionString to metadata manager after server starts. Required for dynamically bound ports.

Hangleton committed:
- Add visitor to traverse locally emulated remote storage to provide support for test assertions.
- Pass connectionString to metadata manager after server starts. Required for dynamically bound ports.

Hangleton committed:
Add visitTopicPartition(), visitSegment() to the local storage traverser.

Hangleton committed:
Add visitTopicPartition(), visitSegment() to the local storage traverser.

Hangleton committed:
1) Add local remote storage listener to allow tests to be notified of modifications in the storage; 2) Add snaphost support for the local remote storage.

Hangleton committed:
1) Add local remote storage listener to allow tests to be notified of modifications in the storage; 2) Add snaphost support for the local remote storage.

Hangleton committed:
- Add waiter on local remote storage to allow tests to formulate exceptations on asynchronously populated remote storage.
- Set readOffset's initial value to -1 to handle the edge case of a single record for the first segment of a topic-partition.

Hangleton committed:
- Add waiter on local remote storage to allow tests to formulate exceptations on asynchronously populated remote storage.
- Set readOffset's initial value to -1 to handle the edge case of a single record for the first segment of a topic-partition.

satishd committed:
Added config for retention policy for remote log metadata topic and other minor fixes.

satishd committed:
Added config for retention policy for remote log metadata topic and other minor fixes.

Hangleton committed:
Rename LocalRemoteStorage to LocalTieredStorage.

Hangleton committed:
Rename LocalRemoteStorage to LocalTieredStorage.

satishd committed:
Added more logging in RLMMWithTopicStorage.

satishd committed:
Added more logging in RLMMWithTopicStorage.

Hangleton committed:
Use the index size instead of log segment size to enable the creation of a single-record log segment.

Hangleton committed:
Use the index size instead of log segment size to enable the creation of a single-record log segment.

Hangleton committed:
Update createAdmin/Producer/Consumer to take the server endpoint, used when listener port is dynamically bound.

Hangleton committed:
Update createAdmin/Producer/Consumer to take the server endpoint, used when listener port is dynamically bound.

Hangleton committed:
Use initialize(.) in onServerStarted(.)

Hangleton committed:
Use initialize(.) in onServerStarted(.)

Hangleton committed:
Re-order method post-merge from main branch.

Hangleton committed:
Re-order method post-merge from main branch.

Hangleton committed:
Refactor LocalTieredStorage to extract file-related components. Provide the full set of files (segment and indexes) to the visitor to allow tests to write exhaustive assertions.

satishd commented:
@Hangleton Can you reabse your PR?

Hangleton committed:
Add Javadoc and consistency check on RemoteLogFileset.

Hangleton committed:
Add methods in LocalTieredStorageSnapshot to help writing test assertions.

Hangleton commented:
> @Hangleton Can you reabse your PR?

Rebased against `tiered-storage`.

Hangleton committed:
Additional Javadoc for LocalTieredStorage.

Hangleton committed:
Refactor TieredStorageTest. Add TieredStorageTestCase to formulate simple test cases excercising the local tiered storage.

Hangleton committed:
Added another segment to be offloaded in base integration test. Couple of fixes in test assertions.

Hangleton committed:
Add a test case for a segment with multiple records.

Hangleton committed:
Refactor LocalTieredStorageTestCase to make it more readable.

Hangleton committed:
Describe configuration on delete inactive log segments.

satishd commented on clients/src/main/java/org/apache/kafka/common/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -212,6 +212,10 @@ public void syncCommittedDataAndOffsets(boolean forceSync) {
     public void reassignForPartitions(Set<TopicPartition> partitions) {
         Objects.requireNonNull(partitions, "partitions can not be null");
 
+        if (partitions.isEmpty()) {
```
We should not ignore partitions being empty. This case is possible when partitions are reassigned to other brokers and this broker may not have any partitions deployed but it may be assigned with a set of partitions later.

satishd commented on clients/src/main/java/org/apache/kafka/common/log/remote/metadata/storage/RLMMWithTopicStorage.java:
```diff
@@ -121,6 +121,16 @@ private void publishMessageToPartition(RemoteLogSegmentMetadata remoteLogSegment
+        //          segment to be deleted when the metadata was not refreshed and had creationTimestamp == 0,
+        //          which made the remote segment eligible to delete in RLM#handleExpiredRemoteLogSegments().
+        //
+        updateRemoteLogSegmentMetadata(
```
We should not be updating here as it may cause race conditions. All the updates should go through the topic to maintain the order. Segments will be deleted only for segments that have been successfully copied to remote storage.

satishd commented on clients/src/main/java/org/apache/kafka/common/log/remote/storage/RemoteLogMetadataManager.java:
```diff
@@ -139,5 +139,5 @@
      * server is started.
      */
-    void onServerStarted();
+    void onServerStarted(final Optional<String> serverEndpoint);
```
We can leave Optional here, it is not intended to be used as a method arg but it can be a good choice for return type. Good discussion SO is [here](https://stackoverflow.com/questions/31922866/why-should-java-8s-optional-not-be-used-in-arguments).

satishd commented on clients/src/main/java/org/apache/kafka/common/log/remote/metadata/storage/RLMMWithTopicStorage.java:
```diff
@@ -254,17 +264,17 @@ public void onStopPartitions(Set<TopicPartition> partitions) {
-            createAdminClient();
-            createProducer();
-            createConsumer();
+            createAdminClient(endpoint);
```
I put a comment in `RemoteLogMetadataManager` to remove Optional wrapper. 
We should throw an error(possibly `IllegalArgumentException`) if endpoint is null as it should always exist for the broker and it is needed to create the clients.

satishd commented on clients/src/test/java/org/apache/kafka/common/log/remote/storage/LocalTieredStorage.java:
```diff
@@ -16,48 +16,66 @@
-    private static final Logger LOGGER = LoggerFactory.getLogger(LocalRemoteStorageManager.class);
+    private static final Logger LOGGER = LoggerFactory.getLogger(LocalTieredStorage.class);
 
     /**
```
Good to push these docs to the respective properties defined earlier. This will make it easy for users to understand them and set the appropriate values.

satishd commented on clients/src/test/java/org/apache/kafka/common/log/remote/storage/LocalTieredStorage.java:
```diff
@@ -112,15 +169,20 @@ public void configure(Map<String, ?> configs) {
 
-        Directory directory = openDirectory(ROOT_STORAGES_DIR_NAME + "/" + storageId, new File("."));
-        storageDirectory = directory.directory;
+        storageDirectory = new File(new File("."), ROOT_STORAGES_DIR_NAME + "/" + storageId);
```
Good to have a config for parent directory, it should default to a temp dir if it is not configured.

satishd reviewed (commented):
thanks @Hangleton for the PR, nice work on tests.
Left a few comments which need to be addressed.

Hangleton committed:
Remove hacks in remote log metadata manager.
Remove Optional for the server endpoint provided at initialization of the remote log metadata manager.
Create a temporary directory in LocalTieredStorage if no directory is defined in configuration properties.

Hangleton committed:
Fix for remote log metatdata manager integration test.

Hangleton committed:
Move documentation to property level in LocalTieredStorage.

Hangleton committed:
Fix build.

satishd commented:
Thanks @Hangleton for addressing the review comments. We will have a followup PR for pending comments.
